[WIP] Created __repr__ functions for learners + preprocessors

Ameobea opened the pull request:
This commit creates `__repr__` functions for the majority of Orange's learner and preprocessor classes.  It uses standardized and generic functions wherever possible.  These changes pave the way for my code generator (PR upcoming), which makes extensive use of these functions to create runnable Python scripts based on the contents of the workflow.

I moved some data processing functions out of widget code and into more appropriate directories, updating the widget code as appropriate.  I converted some functions to classes in order to make them more similar to others and created **repr** functions for them as well.  Their usages were updated accordingly.  I then squashed all commits into a single commit.

All tests are passing (except those that weren't passing before) and the linter doesn't show anything significant caused by this commit.  As far as I can tell, this commit is ready to merge and completely working.  If any changes need to be made, please let me know.

EDIT: I've changed the title to WIP due to issues between the results of tests on my computer and those returned by Travis.

Ameobea commented:
I'm sorry for all the commits; when I run the tests on my computer, the results are completely different than when they are run by Travis.

kernc commented:
Don't worry. It seems [this line](https://github.com/biolab/orange3/pull/1507/files#diff-cd2f62cf361dc130ed42eb2ee10236c1R198) in _owdatasampler.py_ breaks due to `self.indices` no longer being iterable, whereas it was before. You can run the same tests that Travis does by running [this command](https://github.com/biolab/orange3/blob/master/.travis/stage_script.sh#L14).

Ameobea commented:
I'm going to write some tests for the new `__repr__` functions and update this PR.

Ameobea commented:
I've created tests for every Preprocessor, Filter, and Learner impacted by this PR.  Code coverage remains low, but I can't find any new blocks of code that I've written that aren't covered by tests.

If somebody has a recommendation that would improve the code coverage or other improvement in general, please let me know.  However, from everything I can see, this PR is tested and ready to merge, so I will be removing the [WIP] from the title.

astaric commented:
The main cause of the low coverage is the Scaling preprocessor. Could you include some tests for it?

Ameobea commented:
Ahh I totally forgot that I included that; it was copied over from the `owpreprocess` widget.  I'll write tests for that first thing tomorrow.

astaric commented on Orange/tests/test_preprocess.py:
```diff
@@ -83,6 +84,15 @@ def test_remove_nan_classes_multiclass(self):
+    def test_scaling_x(self):
+        table = Table("iris")
+        dist1 = distribution.get_distribution(table, table.domain.attributes[0])
+        self.assertTrue(Scaling.mean(dist1) > 5)
```
We prefer to use self.assertGreater, since the error it prouduces is more informative. Also, could you include the checks for median, stdev and span as well (they could be done in two pairs, mean&stdev and median&span)?

This should increase the coverage enough to make the codecov happy.

kernc commented on Orange/preprocess/discretize.py:
```diff
@@ -98,6 +98,14 @@ def __call__(self, data, variable):
+            self.__class__.__name__,
+            ", ".join("{}={}".format(arg, repr(getattr(self, arg))) for i, arg in enumerate(args) if
+                arg != "self" and self.__class__.__init__.__defaults__[i-1] != getattr(self, arg))
+        )
```
Put this in a separate class:
```py
class Reprable:
    def __repr__(self):
        """ Docstring explaining what happens """
        ...
```
Then you can have `Discretization` and `Preprocess` (below) inherit this new type?

kernc commented on Orange/preprocess/impute.py:
```diff
@@ -69,6 +69,9 @@ class DoNotImpute(BaseImputeMethod):
         return variable
 
+    def __repr__(self):
+        return "DoNotImpute()"
```
Better define a generic on `BaseImputeMethod`?
```py
def __repr__(self):
    return type(self).__name__ + '()'
```

Ameobea replied:
Don't know why I didn't do that to begin with...

kernc commented on Orange/preprocess/impute.py:
```diff
@@ -129,6 +141,9 @@ def __call__(self, data, variable, *, default=None):
         return Default(self.default)
 
+    def __repr__(self):
+        return "Default()"
```
`Default` constructor accepts arguments and this `__repr__ doesn't account for that (same with `ImputeSql` above). Could the [`Reprable` idea](https://github.com/biolab/orange3/pull/1507/files#r78161092) be used here as well?

kernc replied:
`Default` constructor accepts arguments and this `__repr__` doesn't account for that (same with `ImputeSql` above). Could the [`Reprable` idea](https://github.com/biolab/orange3/pull/1507/files#r78161092) be used here as well?

kernc commented on Orange/preprocess/preprocess.py:
```diff
@@ -422,3 +504,9 @@ def __call__(self, data):
+        for preproc in self.preprocessors:
+            repstr +=  repr(preproc) + ", "
+        repstr += "])"
+        return repstr
```
Since 
```py
repr(self.preprocessors) == ', '.join(repr(i) for i in self.preprocessors)
```
this could also work with [`Reprable`](https://github.com/biolab/orange3/pull/1507/files#r78161092), right?

kernc commented on Orange/preprocess/score.py:
```diff
@@ -58,6 +58,14 @@ def __call__(self, data, feature=None):
+            self.__class__.__name__,
+            ", ".join("{}={}".format(arg, repr(getattr(self, arg))) for i, arg in enumerate(args) if
+                arg != "self" and self.__class__.__init__.__defaults__[i-1] != getattr(self, arg))
+        )
```
This is, again, something `Reprable` would unify in a single place. Note, `arg != 'self'` check can be avoided with:
```py
args = type(self).__init__.__code__.co_varnames[1:]
```

kernc commented on Orange/widgets/data/owcontinuize.py:
```diff
@@ -378,6 +378,18 @@ def __call__(self, data):
+            "continuous_treatment=Continuize.{}, ".format(repr(self.continuous_treatment)) if
+                self.continuous_treatment != Continuize.Leave else "",
+            "class_treatment=Continuize.{}, ".format(repr(self.class_treatment)) if
+                self.class_treatment != Continuize.Leave else ""
```
I think this might work with `Reprable` also. The only botherseome thing, on multiple counts, is the use of `__new__`. I'm totally _for_ stripping `data` arg from `__new__` (now `__init__`) and then replacing all occurrences of `DomainContinuizer(data, ...)` with `DomainContinuizer(...)(args)`, i.e. construct the continuizer object first, then call it to transform the data, like the rest of preprocessors do.

kernc commented on Orange/widgets/data/owdatasampler.py:
```diff
@@ -294,62 +293,100 @@ def send_report(self):
-
-
-def sample_random_n(table, n, stratified=False, replace=False,
+class sample_fold_indices():
```
Class' name should be `CamelCase`.

kernc commented on Orange/widgets/data/owdatasampler.py:
```diff
@@ -294,62 +293,100 @@ def send_report(self):
+        :param int folds: Number of folds
+        :param bool stratified: Return stratified indices (if applicable).
+        :param Random random_state:
+        :rval tuple-of-arrays: A tuple of array indices one for each fold.
```
Since this is new code, could you also make this a [NumPy or Google style docstring](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/index.html#google-vs-numpy)?

kernc commented on Orange/widgets/data/owdatasampler.py:
```diff
@@ -294,62 +293,100 @@ def send_report(self):
+        return "{}({})".format(
+            self.__class__.__name__,
+            ", ".join("{}={}".format(arg, repr(getattr(self, arg))) for i, arg in enumerate(args) if
+                arg != "self" and self.__class__.__init__.__defaults__[i-1] != getattr(self, arg))
```
`Reprable`.

kernc commented:
This PR is certainly a step in the right direction! Thanks for working on it. :+1:

Ameobea commented:
I've made changes accounting for all of the comments above.  Thank you for having the patience to deal with my ineptitude for this particular aspect of the project.  Since I've mostly worked only on my own projects in the past, none of my work has included stuff like automated testing, complicated inheritance, or issues with circular dependencies with completely generic error messages :unamused:

Anyway, thanks again for all the help; I look forward to continuing work and making the project as good as it can be.

kernc commented on Orange/misc/enum.py:
```diff
@@ -73,9 +73,7 @@ class EnumValue(int):
-            return str(names[self])
-
-        __str__ = __repr__
+            return self.__str__()
```
Are you sure this is still the same? A string was returned before, I think.

kernc commented on Orange/preprocess/discretize.py:
```diff
@@ -98,6 +99,14 @@ def __call__(self, data, variable):
             "Subclasses of 'Discretization' need to implement "
             "the call operator")
 
+    def __repr__(self):
```
So now I guess you can remove this override.

kernc commented on Orange/preprocess/fss.py:
```diff
@@ -89,6 +88,15 @@ def __call__(self, data):
                                     data.domain.class_vars, data.domain.metas)
         return data.from_table(domain, data)
 
+    def __repr__(self):
```
Everywhere where extending `Reprable` would work.

kernc commented on Orange/preprocess/fss.py:
```diff
@@ -127,6 +135,8 @@ def __call__(self, data):
             data.domain.class_vars, data.domain.metas)
         return data.from_table(domain, data)
 
+    __repr__ = SelectBestFeatures.__repr__
```
`Reprable`.

kernc commented on Orange/preprocess/reprable.py:
```diff
@@ -0,0 +1,39 @@
+class Reprable:
+    def __repr__(self):
+        """ A helper class that, when inherited, creates an executable __repr__
+        String based on the arguments passed to the preprocessor's __init__ function.
```
I wouldn't say this is necessarily for preprocessors only. Approximately the same code can be used for nearly arbitrary objects we have, including `Learner` and `Model` objects?

kernc commented on Orange/preprocess/reprable.py:
```diff
@@ -0,0 +1,39 @@
+            return "{}({})".format(
+                self.__class__.__name__,
+                ", ".join("{}={}".format(arg, repr(getattr(self, arg))) for i, arg in enumerate(args) if
+                    self.__class__.__init__.__defaults__[i-1] != getattr(self, arg))
```
Can you break this a little nicer:
```py
', '.join(something
          for something in something_else
          if (certain condtion and
              certain other condition))
```
Please also write in a few English words what is actually happening there.

kernc replied:
`self.__class__.__init__.__defaults__[i-1]` for `i=0` is `self.__class__.__init__.__defaults__[-1]`?

kernc commented on Orange/preprocess/reprable.py:
```diff
@@ -0,0 +1,39 @@
+            )
+
+
+class CallReprable:
```
Doesn't appear to be used anywhere in the PR?

In either case, it's so similar to `Reprable` that if needed, it had better be parameterized, e.g.
```py
_reprable_params_method = '__call__'  # And then a single Reprable can getattr on that
```

kernc replied:
But this should not be needed at all. One doesn't _construct_ objects by invoking their `__call__`.

kernc commented on Orange/preprocess/reprable.py:
```diff
@@ -0,0 +1,39 @@
+class Reprable:
```
As said, I would move this out of _preprocess_ as its applications are much wider.

kernc commented on Orange/tests/test_classification.py:
```diff
@@ -313,3 +319,28 @@ def test_adequacy_all_learners_multiclass(self):
+        for l in learners:
+            repr_str = repr(l)
+            new_l = eval(repr_str)
+            self.assertEqual(repr(new_l), repr_str)
```
So this is basically
```py
for learner in learners:
    self.assertEqual(repr(learner), repr(eval(repr(learner))))
```
The thing this code doesn't check is equality of `repr(learner)` to the above instantiation line (e.g. `LogisticRegressionLearner(tol=0.0002)`). Could these instantiantions be in string form and eval'd first.

kernc commented on Orange/base.py:
```diff
@@ -69,7 +69,11 @@ def preprocess(self, data):
+            ", ".join("{}={}".format(k, repr(v))
+                for k, v in self.params.items()) if
+                "params" in dir(self) else ""
+       )
```
Can you make `Learner` and `Model` work with `Reprable` too? For the moment, checking `getattr(self, arg)` or `self.params[arg]` should do the job.

astaric commented:
Regarding the stalled travis build, it seems to me that travis will not run the tests until the code can be cleanly merged into the latest master. Could you rebase?

Ameobea commented:
I've converted the global repr generator to a much more robust format using `inspect.getargspec`.  I've moved the `Reprable` class to the misc director out of the preprocessor directory and implemented it for Filters.

I have more work to do before seeking approval for a merge, so I've (yet again) re-set the WIP tag for this PR.  I will work soon to implement the global repr function for all internal orange objects, write additional tests, and make sure everything is polished and high-quality before seeking a merge.

kernc commented:
That's great! :+1:   Just consider `inspect.getargspec()` is [deprecated](https://docs.python.org/3/library/inspect.html#inspect.getargspec).

astaric commented:
Hey, it has been a month since the last activity on this PR. IIUC, this used to be a summer project, so please let us know if you do not intend to finish the feature, so we can either close the pull request (you can always reopen it if you decide to complete the work in the future) or have someone else try to finish the implementation. If you are still working on this, you will need to rebase the code onto the latest master, as it currently cannot be merged cleanly.

kernc commented:
Took over in https://github.com/biolab/orange3/pull/1771.

@Ameobea hope you don't mind me squashing them big commits. Thanks for your work on this; it's gonna be a blast!
